Misc developments for full Run 2

ftenchini opened the pull request:
- Allows wremnant to run on new Pisa machine (default path is set for 2016, 2018 paths are relative to it)
- Added custom paths for BKG samples
- Improved the trigger object matching criteria 

Checked by running mw, mz-dilepton, and mz-wlike histomakers for 2016 & 2018.
Also checked mw histomaker with `--makeMCefficiency` option running on WplusMunu sample for 2016 & 2018.

cippy commented:
It seems the CI must be approved to run, apparently because Francesco doesn't appear in the list of contributors, and I am not sure he would need additional permission in general for this repository. 

I will click on "Approve and run", and based on the documentation next time no further action should be needed from him to start CI automatically.

bendavid commented:
Yes this is the intended behaviour.  In principle a random malicious person could try to run harmful code through the ci so we have this first-time approval which is needed.

davidwalter2 commented on scripts/histmakers/mw_with_mu_eta_pt.py:
```diff
@@ -290,10 +290,11 @@ def build_graph(df, dataset):
     if args.makeMCefficiency:
-        df = df.Define("GoodTrigObjs", "wrem::goodMuonTriggerCandidate(TrigObj_id,TrigObj_filterBits)")
-        df = df.Define("passTrigger","(HLT_IsoTkMu24 || HLT_IsoMu24) && wrem::hasTriggerMatch(goodMuons_eta0,goodMuons_phi0,TrigObj_eta[GoodTrigObjs],TrigObj_phi[GoodTrigObjs])")
+        df = df.Define("GoodTrigObjs", f"wrem::goodMuonTriggerCandidate(TrigObj_id,TrigObj_filterBits, \"{era}\")")
```
The trigger that is used is known at runtime, I suggest to solve this a bit different. For example, defining two separate functions for 2016 and 2017/2018.

cippy replied:
Hi David, thanks for the suggestion. Originally we were actually doing as you say, but some functions are used internally by others, for example here (where by the way you may have spotted a bug) [0]. So we might still have to allow for an "era" flag in the function definition sometimes. At that point we thought it was simpler to have it wherever it could be used, keeping 2016 as default to minimally affect other pieces of code. 

In case you were also concerned about the usage of strings for the era, we might define an enum class paired with a dictionary to map integers into string for printing purpose, similarly to what is currently done for the analysis type, so to pass an integer type to all functions.

[0] https://github.com/WMass/WRemnants/blob/a9276858cf93cbd97188ac11cabd8173dc28e7ea/wremnants/muon_selections.py#L100C5-L109

davidwalter2 replied:
I think it's fine to have the era flag in python functions, I would try to avoid it in the c++ function because those are then used in the computational graph. Another possibility is to write c++ template functions, but two different c++ functions would be simpler in my opinion.

ftenchini replied:
I would not use distinct functions for each era because we would end up with duplicated code.

It _does_ make sense to have a single function of `era` (and other parameters if needed) that assembles and outputs the trigger bit string. 

I would keep things as is right now because it's only a single if statement. If the trigger definition becomes more complex then we can replace it with a function.

davidwalter2 replied:
To avoid duplication of code you can make it a template function, something like
```
enum class Era {
    Era_2016PostVFP,
    // Add other eras as needed
};

template <Era era>
Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_f& TrigObj_pt, const Vec_f& TrigObj_l1pt, const Vec_f& TrigObj_l2pt, const Vec_i& TrigObj_filterBits) {
    Vec_i res(TrigObj_id.size(), 0);
    if (TrigObj_id[i]  != 13 ) continue;
    if (TrigObj_pt[i]   < 24.) continue;
    if (TrigObj_l1pt[i] < 22.) continue;
    if constexpr (era == Era::Era_2016PostVFP){
      if (! ((TrigObj_filterBits[i] & 8) || (TrigObj_l2pt[i] > 10. && (TrigObj_filterBits[i] & 2))))) continue;
    }
    if constexpr (era != Era::Era_2016PostVFP){
      if (! (TrigObj_l2pt[i] > 10. && (TrigObj_filterBits[i] & 2)))) continue;
    }
    res[i] = 1;
   }
   return res;
}
```
and call the function as 
```
goodMuonTriggerCandidate<Era::Era_2016PostVFP>(TrigObj_id, TrigObj_pt, TrigObj_l1pt, TrigObj_l2pt, TrigObj_filterBits);
```
This make the graph indepent of era avoids passing the era string into the function for every single event

cippy replied:
I would be in favour of this (either the template or the duplication), mainly because I suspect using string comparison per-event can be quite slow (and unnecessary).
Since currently we foresee only two cases, instead of two independent "if"s we can use "if .. else".

ftenchini replied:
I also agree. I'll work on this.

davidwalter2 commented on wremnants/muon_selections.py:
```diff
@@ -97,7 +97,7 @@ def select_z_candidate(df, mass_min=60, mass_max=120):
     return df
 
-def apply_triggermatching_muon(df, dataset, muon_eta, muon_phi, otherMuon_eta=None, otherMuon_phi=None):
+def apply_triggermatching_muon(df, dataset, muon_eta, muon_phi, otherMuon_eta=None, otherMuon_phi=None, era = "2016PostVFP"):
```
I think era must be used somewhere here to get the right trigger

ftenchini replied:
Should be fixed now.

davidwalter2 commented on wremnants/include/utils.h:
```diff
@@ -134,12 +135,13 @@ Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_f& TrigObj_pt,
 
 // new overloaded function to be used with new ntuples having additional trigger bits
-Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_i& TrigObj_filterBits) {
+Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_i& TrigObj_filterBits, const string& era = "2016PostVFP") {
```
I think separate functions would be better

davidwalter2 commented on wremnants/include/utils.h:
```diff
@@ -118,14 +118,15 @@ Vec_i cleanJetsFromLeptons(const Vec_f& Jet_eta, const Vec_f& Jet_phi, const Vec
     
-Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_f& TrigObj_pt, const Vec_f& TrigObj_l1pt, const Vec_f& TrigObj_l2pt, const Vec_i& TrigObj_filterBits) {
-
+Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_f& TrigObj_pt, const Vec_f& TrigObj_l1pt, const Vec_f& TrigObj_l2pt, const Vec_i& TrigObj_filterBits, const string&  era = "2016PostVFP") {
```
I think separate functions would be better

ftenchini commented:
Added `enum class Era`.
Turned `goodMuonTriggerCandidate` into a template function.
Moved the construction of the trigger bit string to its own function. 
Fixed a couple of bugs.
